File: src/javascript/xpath-injection.js
// CWE-643: XPath injection.
// Rule: js/xpath-injection
// Autofix: typically NOT provided — XPath has no parameterized API in
// xpath/xmldom; fix requires either an escaping function or a redesign
// using XPath variables (xpath.parse + evaluate with context).

const express = require('express');
const xpath = require('xpath');
const { DOMParser } = require('xmldom');

const app = express();
const doc = new DOMParser().parseFromString(`
  <users>
    <user role="admin"><name>alice</name><email>a@example.com</email></user>
    <user role="user"><name>bob</name><email>b@example.com</email></user>
  </users>
`);

app.get('/user', (req, res) => {
  const name = String(req.query.name || '');
  const expr = `//user[name="${name}"]/email/text()`;
  const nodes = xpath.select(expr, doc);
  res.json(nodes.map(n => n.toString()));
});

app.get('/by-role', (req, res) => {
  const role = String(req.query.role || '');
  const nodes = xpath.select(`//user[@role='${role}']/name/text()`, doc);
  res.json(nodes.map(n => n.toString()));
});

app.listen(3106);

File: src/python/xpath_injection.py
"""CWE-643: XPath injection.

Rule: py/xpath-injection
Autofix: typically NOT provided — fix requires switching to `lxml.etree.XPath`
with variables, or escaping; needs API redesign.
"""

from lxml import etree
from flask import Flask, request

app = Flask(__name__)

XML = etree.fromstring(b"""
<users>
  <user role="admin"><name>alice</name><email>a@example.com</email></user>
  <user role="user"><name>bob</name><email>b@example.com</email></user>
</users>
""")


@app.route("/email")
def email():
    name = request.args.get("name", "")
    nodes = XML.xpath(f"//user[name='{name}']/email/text()")
    return {"emails": list(nodes)}


@app.route("/by-role")
def by_role():
    role = request.args.get("role", "")
    nodes = XML.xpath('//user[@role="' + role + '"]/name/text()')
    return {"names": list(nodes)}


if __name__ == "__main__":
    app.run(port=4106)
